File: source/downloads/code/LeetCode/Java/Path-Sum2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    
    List<List<Integer>> ret = new LinkedList<List<Integer>>();
    List<Integer> tmp = new LinkedList<Integer>();
 
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        if (root == null)
            return ret;
        dfs(root, sum);
        return ret;       
    }

    public void dfs(TreeNode root, int sum){
        tmp.add(root.val);
        if (sum == root.val && root.left==null && root.right==null) {
            ret.add(new LinkedList(tmp));       
            return ;    
        }
        
        if (root.left != null){
            dfs(root.left, sum-root.val);
            tmp.remove(tmp.size()-1);
        }
        if (root.right != null){
            dfs(root.right, sum-root.val);
            tmp.remove(tmp.size()-1);
        }
        
    }
}

File: downloads/code/LeetCode/Python/Path-Sum2.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):

    def findPath(self,result,vector,node,bal):
        vector.append(node.val)
        if node.val == bal and node.left == None and node.right == None:
            result.append(vector[:])
            return

        if node.left != None:
            self.findPath(result,vector,node.left,bal-node.val)
            vector.pop()
        if node.right != None:
            self.findPath(result,vector,node.right,bal-node.val)
            vector.pop()
        return




    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
    
        result = []
        vector = []
        if root == None:
            return result

        self.findPath(result,vector,root,sum);
        return result
